src/Components/Components/src/Environment/ComponentEnvironment.cs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Environment
+{
+    /// <summary>
+    /// Provides information an environment specific features about the environment the components are running on.
+    /// </summary>
+    public class ComponentEnvironment
+    {
+        /// <summary>
+        /// Name of associated with the <see cref="ComponentEnvironment"/> when components are being remotely rendered on the server.
+        /// </summary>
+        public const string Remote = nameof(Remote);
+
+        /// <summary>
+        /// Name of associated with the <see cref="ComponentEnvironment"/> when components are being prerrendered on the server.
+        /// </summary>
+        public const string Prerrender = nameof(Prerrender);
+
+        /// <summary>
+        /// Name of associated with the <see cref="ComponentEnvironment"/> when components are being locally rendered on the client.
+        /// </summary>
+        public const string Local = nameof(Local);
+
+        /// <summary>
+        /// Gets or sets identifier associated with the <see cref="ComponentEnvironment"/> the components are running on.
+        /// </summary>
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="IJSRuntime"/> for this <see cref="ComponentEnvironment"/>.
+        /// </summary>
+        public IJSRuntime JSRuntime { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="IUriHelper"/> for this <see cref="ComponentEnvironment"/>.
+        /// </summary>
+        public IUriHelper UriHelper { get; set; }
```

Comment:
I really don't think we should put services available from DI onto a context object. Anyone who needs these things should get them from DI.

Reply:
You can still get them through DI, but we need a way to provide different implementations whether you are rendering components on a prerrendering environment or on a non prerrendered environment, as they require different behaviors.

Reply:
If they are defined as non-singletons (presumably, scoped), can't the DI system conditionally return different values depending on whether we're in a prerendering context or not?

Reply:
BTW I've nothing against the existence of `ComponentEnvironment` itself. It's certainly a nice way for components to find out whether they are currently prerendering or not. If we could make other services injectable independently of it, that would be great.